Panic when publishing

I saw this by accident and have no idea really why. Just thought I would drop it here.

``` bash
panic: sync: WaitGroup is reused before previous Wait has returned

goroutine 249599 [running]:
panic(0x7eba60, 0xc83d589f70)
    /usr/local/go/src/runtime/panic.go:481 +0x3e6
sync.(*WaitGroup).Wait(0xc8200598e8)
    /usr/local/go/src/sync/waitgroup.go:129 +0x114
github.com/nats-io/nats.(*Conn).waitForExits(0xc820059800)
    /data/jenkins/jobs/tgm_ng deploy/workspace/vendor/src/github.com/nats-io/nats/nats.go:733 +0x6a
github.com/nats-io/nats.(*Conn).doReconnect(0xc820059800)
    /data/jenkins/jobs/tgm_ng deploy/workspace/vendor/src/github.com/nats-io/nats/nats.go:1084 +0x25
created by github.com/nats-io/nats.(*Conn).processOpErr
    /data/jenkins/jobs/tgm_ng deploy/workspace/vendor/src/github.com/nats-io/nats/nats.go:1232 +0x25f
```

New report from Google Groups:
encountered a panic in one of my NATS clients:

```
panic: sync: WaitGroup is reused before previous Wait has returned

goroutine 372091 [running]:
sync.(*WaitGroup).Wait(0xc420499328)
       /usr/local/go/src/sync/waitgroup.go:133 +0xd8
github.com/nats-io/go-nats.(*Conn).waitForExits(0xc420499200)
       /XXX/src/github.com/nats-io/go-nats/nats.go:825 +0x6a
github.com/nats-io/go-nats.(*Conn).spinUpGoRoutines(0xc420499200)
       /XXX/src/github.com/nats-io/go-nats/nats.go:835 +0x2f
created by github.com/nats-io/go-nats.(*Conn).processConnectInit
       /XXX/github.com/nats-io/go-nats/nats.go:919 +0x1a9
```

It seems like I have had the same problem described here (almost 10 months old, ticket closed): https://github.com/nats-io/go-nats/issues/192. I don't have the full stack that occurred either.

I believe this is related to my STAN server being restarted and also that my client is configured to reconnect infinitely, but I can't be sure without full trace:

```
config := nats.Options{
  MaxReconnect: -1,
}
```

My questions:
Is this a really rare panic ? I see the github ticket is closed, but yet I encountered the panic. Is there anything to do to avoid that panic ?
Is MaxReconnect: -1 the best way to have infinite reconnect try ?

--- a/nats.go
+++ b/nats.go
@@ -246,7 +246,7 @@ type Conn struct {
 	Statistics
 	mu      sync.Mutex
 	Opts    Options
-	wg      sync.WaitGroup
+	wg      *sync.WaitGroup
 	url     *url.URL
 	conn    net.Conn
 	srvPool []*srv
@@ -872,7 +872,7 @@ func (nc *Conn) makeTLSConn() {
 
 // waitForExits will wait for all socket watcher Go routines to
 // be shutdown before proceeding.
-func (nc *Conn) waitForExits() {
+func (nc *Conn) waitForExits(wg *sync.WaitGroup) {
 	// Kick old flusher forcefully.
 	select {
 	case nc.fch <- true:
@@ -880,7 +880,9 @@ func (nc *Conn) waitForExits() {
 	}
 
 	// Wait for any previous go routines.
-	nc.wg.Wait()
+	if wg != nil {
+		wg.Wait()
+	}
 }
 
 // spinUpGoRoutines will launch the Go routines responsible for
@@ -890,14 +892,16 @@ func (nc *Conn) waitForExits() {
 // reconnect when the previous ones have exited.
 func (nc *Conn) spinUpGoRoutines() {
 	// Make sure everything has exited.
-	nc.waitForExits()
+	nc.waitForExits(nc.wg)
 
+	// Create a new waitGroup instance for this run.
+	nc.wg = &sync.WaitGroup{}
 	// We will wait on both.
 	nc.wg.Add(2)
 
 	// Spin up the readLoop and the socket flusher.
-	go nc.readLoop()
-	go nc.flusher()
+	go nc.readLoop(nc.wg)
+	go nc.flusher(nc.wg)
 
 	nc.mu.Lock()
 	if nc.Opts.PingInterval > 0 {
@@ -1232,7 +1236,10 @@ func (nc *Conn) flushReconnectPendingItems() {
 func (nc *Conn) doReconnect() {
 	// We want to make sure we have the other watchers shutdown properly
 	// here before we proceed past this point.
-	nc.waitForExits()
+	nc.mu.Lock()
+	wg := nc.wg
+	nc.mu.Unlock()
+	nc.waitForExits(wg)
 
 	// FIXME(dlc) - We have an issue here if we have
 	// outstanding flush points (pongs) and they were not
@@ -1423,9 +1430,9 @@ func (nc *Conn) asyncDispatch() {
 // readLoop() will sit on the socket reading and processing the
 // protocol from the server. It will dispatch appropriately based
 // on the op type.
-func (nc *Conn) readLoop() {
+func (nc *Conn) readLoop(wg *sync.WaitGroup) {
 	// Release the wait group on exit
-	defer nc.wg.Done()
+	defer wg.Done()
 
 	// Create a parseState if needed.
 	nc.mu.Lock()
@@ -1629,9 +1636,9 @@ func (nc *Conn) processPermissionsViolation(err string) {
 
 // flusher is a separate Go routine that will process flush requests for the write
 // bufio. This allows coalescing of writes to the underlying socket.
-func (nc *Conn) flusher() {
+func (nc *Conn) flusher(wg *sync.WaitGroup) {
 	// Release the wait group
-	defer nc.wg.Done()
+	defer wg.Done()
 
 	// snapshot the bw and conn since they can change from underneath of us.
 	nc.mu.Lock()

--- a/staticcheck.ignore
+++ b/staticcheck.ignore
@@ -1,2 +1,3 @@
 github.com/nats-io/go-nats/*_test.go:SA2002
 github.com/nats-io/go-nats/*/*_test.go:SA2002
+github.com/nats-io/go-nats/nats.go:SA6000

